### ISSUE TITLE
explicitAttrs function not working properly?

### ISSUE THREAD

**Issue by SBlechmann** [telefonicaid/iotagent-node-lib#1016]
Hey there,

as statet in #1003 with the new lib version I was able to get the autoprovision function working. 

I also tried to use the function of explicitAttrs. I created two devices, one with explicitAttrs=true and one with false. When sending data via MQTT to the respective devices using non-provisioned attributes both times attributes are created in the context broker.

Am I using the function wrong or is there a bug?

Thanks for your time!

`{
      "device_id": "tsensor001",
       ...[other stuff]...
      "transport": "MQTT",
      "attributes": [
        {
          "object_id": "t",
          "name": "temperature",
          "type": "Number"
        }
      ],
      "lazy": [],
      "commands": [],
      "static_attributes": [],
      "internal_attributes": [],
      "protocol": "PDI-IoTA-JSON",
      "explicitAttrs": true
    },`

**Comment by fgalan** [comment]
It would be great if you could provide the exact sequence of steps to reproduce the problem (curl and mqtt_publish statements are preferred) since you start the agent with an empty database.

**Comment by SBlechmann** [comment]
I'd reference back to #1003.

**Comment by fgalan** [comment]
> I'd reference back to #1003.

Not sure about what you mean... Do you mean the sequence of step is *exactly the same* you describe at https://github.com/telefonicaid/iotagent-node-lib/issues/1003#issuecomment-817155239 ?

**Comment by SBlechmann** [comment]
Yes, I have only conducted the tests described in #1003. 
The results shown there reflect the behaviour of explicitAttrs as well.

**Comment by mapedraza** [comment]
After replicating what @SBlechmann was doing using iotagent-json:1.17.1, I got the following conclusions.

### Using NGSI V1
- with `explicitAttrs`=true on Config group JSON
   - `explicitAttrs`=true on device provision: New attribute is not included  - OK
   - `explicitAttrs`=false on device provision: New attribute is not included - NOK (Device parameter should overwrite group option)
   - `explicitAttrs` not present on device provision: New attribute is not included - OK
- with `explicitAttrs`=false on Config group JSON
   - New attributes are not included 


### Using NGSI V2 (`IOTA_CB_NGSI_VERSION=v2`)
- with `explicitAttrs`=true on Config group JSON
   - `explicitAttrs`=true on device provision: New attribute is included  - NOK
   - `explicitAttrs`=false on device provision: New attribute is included  - OK
   -  `explicitAttrs` not present on device provision: New attribute is included - NOK
- with `explicitAttrs`=false on Config group JSON
   - `explicitAttrs`=true on device provision: New attribute is included  - NOK
   - `explicitAttrs`=false on device provision: New attribute is included  - OK
   -  `explicitAttrs` not present on device provision: New attribute is included - OK

As a resume, using NGSI V1 it seems that always discard new attributes, and using NGSI V2, it seems it always include new attributes coming from device

@SBlechmann , could you confirm this is what you are experiencing?

**Comment by fgalan** [comment]
> As a resume, using NGSI V1 it seems that always discard new attributes, and using NGSI V2, it seems it always include new attributes coming from device

I'd suggest to leave the NGSIv1 case appart. It is a deprecated API and in the process of being removed from the library (see PR https://github.com/telefonicaid/iotagent-node-lib/pull/995). Let's focus on fixing any potential problem of the NGSIv2 case.

**Comment by SBlechmann** [comment]
Sorry for the late reply! Yes, that's what I experienced with regards to v2.

**Comment by fgalan** [comment]
Maybe related issue: https://github.com/telefonicaid/iotagent-node-lib/issues/1045

**Comment by fgalan** [comment]
After merging PR #1046 this issue could have been solved.

@SBlechmann it would be great if you could test it using `:latest` tag from IOTAs from dockerhub.

**Comment by fgalan** [comment]
> @SBlechmann it would be great if you could test it using `:latest` tag from IOTAs from dockerhub.

In particular, these ones:

```
telefonicaiot/iotagent-ul:lastest
telefonicaiot/iotagent-json:lastest
```

**Comment by SBlechmann** [comment]
Thanks so far!
I'm gonna do combined tests when #1003 and #1051 are fixed then :)

**Comment by mapedraza** [comment]
Checking latest version available, `explicitAttrs` works properly in the following scenarios:

## Creating a device group with explicitAttrs = true
 - Posting data without provisioning the device: No new attributes created - OK
 - Provisioning a device with explicitAttrs = false: New atributes created - OK
 - Provisioning a device with explicitAttrs = true: No new attributes created - OK
 - Provisioning a device without defining explicitAttrs: No new attributes created - OK 

 ## Creating a device group with explicitAttrs = false
 - Posting data without provisioning the device: New attributes created - OK
 - Provisioning a device with explicitAttrs = false: New atributes created - OK
 - Provisioning a device with explicitAttrs = true: No new attributes created - OK
 - Provisioning a device without defining explicitAttrs: New attributes created - OK 

  ## Creating a device group without definig explicitAttrs
 - Posting data without provisioning the device: New attributes created - OK
 - Provisioning a device with explicitAttrs = false: New atributes created - OK
 - Provisioning a device with explicitAttrs = true: No new attributes created - OK
 - Provisioning a device without defining explicitAttrs: New attributes created - OK

**Comment by fgalan** [comment]
Based on the @mapedraza positive report and the recent closing of "sister issue" #1045 we are going to close this issue.

@SBlechmann anyway it would be great if you could test it to confirm it works and provide feedback in this issue (use `telefonicaiot/iotagent-json:latest` or `telefonicaiot/iotagent-ul:latest`). If after this feedback we see something is missing, we can reopen the issue. Thanks!

**Comment by SBlechmann** [comment]
Sorry for the long pause. We are currently in the last stage of setting up a test stand that is communicating with our FIWARE based infrastructure. So hopefully I can give some feedback about that project soon. But right there are still other things I have to focus on. 
I will keep you posted!

**Comment by SBlechmann** [comment]
Hey there, 
sorry for the laste reponse but as promised here my feedback. It seems that there is still something not working properly =/

![grafik](https://user-images.githubusercontent.com/51322874/129238205-feba3410-d1da-4a88-9eec-f7d62fb121ea.png)

**Comment by SBlechmann** [comment]
And here another try. The autoprovisioning ( as discussed in #1003 ) seems to work.

![grafik](https://user-images.githubusercontent.com/51322874/129241550-c7b46dcf-0237-4a8f-9998-add4a5b15292.png)

**Comment by mapedraza** [comment]
I just tried again with latest version and I obtain the same results as I described on my [previous comment above](https://github.com/telefonicaid/iotagent-node-lib/issues/1016#issuecomment-852846244)

Could you share your IoTA configuration (if you are using docker, the ENVARs)?

Also, you can use [this link](https://gist.github.com/mapedraza/c54dfda57c1746e60e39be100bc5f302) as a guide to replicate what I am doing to see if it works or not for you.

**Comment by mapedraza** [comment]
FYI: A test with the scenarios in my previous comment was added here: https://github.com/telefonicaid/iotagent-json/pull/592

**Comment by djs0109** [comment]
Hi @mapedraza , I have a question for understanding. What is the actually expected behaviour of `explicitAttrs`? Following is my understanding:
- If `explicitAttrs=true`: unprovisioned attributes will **not be included** in the update request sent to Context Broker.
- If `explicitAttrs=false`: unprovisioned attributes will be **included** in the update (PATCH) request sent to Context Broker. But these attributes will not be automatically created in Context Broker

Is it that right?

**Comment by mapedraza** [comment]
@djs0109 yes, this is the expected behaviour, just let pass attributes explicitly defined as active(provisioned) attributes.

Just one note. The method the IoT uses in order to persist the measures can differ, depending the value set to the config parameter `appendMode` or the environment var `IOTA_APPEND_MODE `, and the `autoprovision` flag into the group config  ([more info here](https://github.com/telefonicaid/iotagent-node-lib/blob/master/doc/installationguide.md))

**Comment by djs0109** [comment]
Thanks for the quick reply! That is a good point.

**Comment by djs0109** [comment]
Hi @mapedraza, in the documentation it says that the `appendMode/IOTA_APPEND_MODE` is overwritten by `autoprovision` flag in group or device provision. However, I find that it is not the case. So I first create a very simple service group:

POST /iot/services
```
{
 "services": [
   {
    "resource": "/iot/json",
    "apikey": "autoprovision",
    "expressionLanguage": "legacy",
    "explicitAttrs": false,
    "autoprovision": true
   }
 ]
}
```
And then I send an anonymous measurement. When I check the logs in Context Broker I found that the `actionType` is `update`

`INFO@2022-11-16T11:14:08.918Z  logTracing.cpp[148]: Request received: POST /v2/op/update, request payload (269 bytes): {"actionType":"update","entities":[{"humidity":{"type":"string","value":111, ... response code: 422`

So  I restart the IoTagent while setting the `IOTA_APPEND_MODE` to `true`, and I send the same measurement again. The `actionType` turn to `append` now.

`INFO@2022-11-16T11:14:34.952Z  logTracing.cpp[148]: Request received: POST /v2/op/update, request payload (269 bytes): {"actionType":"append","entities":[{"humidity":{"type":"string","value":111, ... response code: 204`

I don't know if this behavior is expected or if it just indicates a tiny bug?
